compiler/rustc_symbol_mangling/src/legacy.rs

```diff
@@ -383,14 +384,26 @@ impl<'tcx> Printer<'tcx> for SymbolPrinter<'tcx> {
         &mut self,
         impl_def_id: DefId,
         args: &'tcx [GenericArg<'tcx>],
-        mut self_ty: Ty<'tcx>,
-        mut impl_trait_ref: Option<ty::TraitRef<'tcx>>,
     ) -> Result<(), PrintError> {
-        let mut typing_env = ty::TypingEnv::post_analysis(self.tcx, impl_def_id);
-        if !args.is_empty() {
-            typing_env.param_env =
-                ty::EarlyBinder::bind(typing_env.param_env).instantiate(self.tcx, args);
-        }
+        let self_ty = self.tcx.type_of(impl_def_id);
+        let impl_trait_ref = self.tcx.impl_trait_ref(impl_def_id);
+        let (typing_env, mut self_ty, mut impl_trait_ref) =
+            if self.tcx.generics_of(impl_def_id).count() <= args.len() {
+                (
+                    TypingEnv::fully_monomorphized(),
+                    self_ty.instantiate(self.tcx, args),
+                    impl_trait_ref.map(|impl_trait_ref| impl_trait_ref.instantiate(self.tcx, args)),
+                )
+            } else {
+                // We are probably printing a nested item inside of an impl.
```

Comment:
This happens when we try to mangle a nested function inside of an impl, like:

```
impl<T> Foo for T {
  fn outer() {
    fn inner() {}
  }
}
```

When mangling `inner`, we try to print it like `<T as Foo>::outer::inner`, however since inner doesn't inherit the substitutions from `outer`, we give it fresh identity substs.

I'm kinda of the opinion that we should print **outer item** def paths as plain old `crate::{impl#0}::outer::inner` rather than printing with the trait ref. I think that would break things tho 🤷